### ISSUE TITLE
Add UEFI APIs in `std::os::uefi`

### ISSUE THREAD

**Issue by Ayush1325** [rust-lang/libs-team#87]
# Proposal

## Problem statement

Hello everyone. I have been working on adding Rust std support for UEFI as a part of Google Summer of Code 2022.

This is a proposal to add some UEFI-specific APIs under `std::os::uefi`. It is related to [PR](https://github.com/rust-lang/rust/pull/100316).

## Motivation, use-cases

In the UEFI environment, most of the services/functionality is accessed using System Table, which is passed to the application at its entry. These APIs basically add ways to access the System Table and some of its most common members.

## APIs to add

APIs I propose to add under `std::os::uefi::env`:

1. `pub fn get_system_table() -> NonNull<c_void>`
2. `pub fn get_system_handle() -> NonNull<c_void>`
3. `pub fn init_globals(handle: NonNull<c_void>, system_table: NonNull<c_void>)`

APIs I propose to add under `std::os::uefi::ffi` (Just re-export `std::os::windows::ffi`)

1. trait `OsStrExt`
2. trait `OsStringExt`

## Links and related work

1. [PR](https://github.com/rust-lang/rust/pull/100316)
4. [My Blog](https://www.programmershideaway.xyz/)
5. [Tracking Issue](https://github.com/rust-lang/rust/issues/100499)

**Comment by Ayush1325** [comment]
cc @dvdhrm
`SystemTable`, `BootServices` and `RuntimeServices` are structures defined in the [UEFI Specification](https://uefi.org/sites/default/files/resources/UEFI_Spec_2_9_2021_03_18.pdf). These structures contain pointers to methods and other things which make a UEFI application/driver function. Pointers to these structures are passed to the application at the entry point.

The reason to provide these methods in std is that when using Rust std (where main has a default signature), there needs to be a way for application developers to get access to the `SystemTable` and `SystemHandle` that were passed to the application. So at the very least, the function 1 and 2 need to present in `std` (3 and 4 are just nice to have)

`r-efi` is a specification crate, so it only contains structures and function signatures, not any implementation. Also, since the `SystemTable` and `SystemHandle` are supplied at the image entry point (which is in std) and stored in std, they have to be acquired from `std`.

**Comment by dvdhrm** [comment]
Exposing the `r-efi` types basically pulls in the entire r-efi crate into the public API. I agree that we should avoid this if possible. I have two alternatives (which can actually be combined as well):

1) Make the global functions you proposed return c_void rather than the typed structure-references. This will strip type-safety, but is a very easy way out at minimal cost, IMO.

2) Allow (or even require?) users to define their UEFI entry-point themselves, rather than doing this in std. This will give them access to the system-table and image handle. You can then have a `setup_std()`, or `register_system_table()`, etc., which takes the system-table from the user and keeps it for `std` to retain.

As I see it, the biggest issue is that UEFI does not allow access to the system-table except through your entry-point. This capability-based model conflicts with the global state expected by rust-std. Most of the rust standard-library just does not allow for a user-supplied context object (Maybe it does? I haven't found a nice way.).

I think we also need to consider what happens to rust code linked into a UEFI C application. In this case the entry-point is provided by the C code, and the rust library/helper is just linked into the application. With your proposed solution, you would be unable to use `std` on UEFI from within that rust code, since you do not control the entry-point. By *not* defining the runtime in `std`, but instead accept a user-supplied context (as suggested in 2.), you allow for such use-cases as well.

On a closing note, with `ExitBootServices()` in mind, all this can become quite fragile. But I think it would be safe to argue that rust-std on UEFI requires you not to exit boot-services.

**Comment by Ayush1325** [comment]
>     1. Make the global functions you proposed return c_void rather than the typed structure-references. This will strip type-safety, but is a very easy way out at minimal cost, IMO.

That should work fine. Incidentally, it might be possible just make it return `NonNull<T>`, and leave the `T` to the user.

>     2. Allow (or even require?) users to define their UEFI entry-point themselves, rather than doing this in std. This will give them access to the system-table and image handle. You can then have a `setup_std()`, or `register_system_table()`, etc., which takes the system-table from the user and keeps it for `std` to retain.

This is something I would like to allow as well. Basically, not generate the custom entry point if `no_main` is used. However, I haven't found a way to do that yet. Requiring a custom entry point would break all the tests, so not sure if that is the best approach.

Incidentally, I already have a function to init the global SystemTable and SystemHandle in `std::os::uefi::env` (currently private), which makes it pretty trivial to allow user-generated entry point with full std support.

**Comment by joshtriplett** [comment]
This seems fine to experiment with as an unstable API. (We're not sure if this is the API we want to stabilize, but for nightly it seems fine to experiment.)

I personally would prefer to see these become wrapper newtypes, to make it harder to accidentally pass BootServices to something expecting SystemTable or vice versa. They don't have to be full structure definitions, just an opaque wrapper around `NonNull<c_void>` with a getter and perhaps an unsafe constructor.

For `init_globals`, will that be called automatically by std's initialization code?

**Comment by Ayush1325** [comment]
I think it would be a good idea to have the wrapper type. I am also thinking if I should convert all access to these types to use `read_volatile`.

**Comment by programmerjake** [comment]
`read_volatile` is almost never what you want to use -- it is *not* for when other software might modify values in parallel (different thread or through interrupts) to your code, you want atomics for that. `volatile` is almost exclusively used when accessing a memory-mapped device which doesn't act like normal memory.

if other code can only modify that memory on the same thread as your code, then either use `&Cell<T>` or use raw pointers and `ptr::read`/`ptr::write` or direct field access to `Copy` fields of a raw pointer (do be careful to not call methods that implicitly create a mutable reference to those fields -- that would be UB since other code still holds independent pointers that they presumably can/will write through later). if you can guarantee no uefi methods are run in some scope, shared references are fine as long as their lifetime expired before calling uefi methods that modify that state.

**Comment by Ayush1325** [comment]
Well, the reason I was suggesting volatile is to avoid compiler caching. In C land, that would mean making the pointer volatile (I think?).

**Comment by pitaj** [comment]
Ah okay. In that case, you should probably use a single-threaded CriticalSection Mutex like https://docs.rs/cortex-m/latest/cortex_m/interrupt/struct.Mutex.html
